### PR TITLE
HRRR Availability and Related Metadata

### LINKED ISSUE #584: [BUG] Lack of Availability of HRRR Weather Models Means 3-Date Azimuth Grid Interpolation Fails for GUNW workflow
```
INFO:RAiDER:Extent of the input is (xmin, ymin, xmax, ymax): -78.05, 41.60, -74.40, 43.55
Traceback (most recent call last):
  File "/opt/conda/envs/RAiDER/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/envs/RAiDER/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/conda/envs/RAiDER/lib/python3.10/site-packages/RAiDER/cli/__main__.py", line 46, in <module>
    main()
  File "/opt/conda/envs/RAiDER/lib/python3.10/site-packages/RAiDER/cli/__main__.py", line 42, in main
    process_entry_point.load()()
  File "/opt/conda/envs/RAiDER/lib/python3.10/site-packages/RAiDER/cli/raider.py", line 620, in calcDelaysGUNW
    cube_filenames = calcDelays([path_cfg])
  File "/opt/conda/envs/RAiDER/lib/python3.10/site-packages/RAiDER/cli/raider.py", line 388, in calcDelays
    raise NotImplementedError(f'The {interp_method} with {n} retrieved weather model files was not well posed '
NotImplementedError: The azimuth_time_grid with 1 retrieved weather model files was not well posed for the dela current workflow.
```

is raised from these inputs:

```
{'execution_started': True,
 'expiration_time': '2023-10-25T00:00:00+00:00',
 'job_id': '00312fde-8f89-4c5c-a0d6-ec2df678f316',
 'job_parameters': {'compute_solid_earth_tide': True,
                    'estimate_ionosphere_delay': True,
                    'frame_id': 16421,
                    'granules': ['S1A_IW_SLC__1SDV_20170210T225922_20170210T225949_015228_018F0A_787A',
                                 'S1A_IW_SLC__1SDV_20170210T225947_20170210T230014_015228_018F0A_10FF'],
                    'secondary_granules': ['S1A_IW_SLC__1SDV_20170129T225922_20170129T225949_015053_018996_BEDD',
                                           'S1A_IW_SLC__1SDV_20170129T225947_20170129T230014_015053_018996_6EA8'],
                    'weather_model': 'HRRR'},
 'job_type': 'INSAR_ISCE_TEST',
 'logs': ['https://hyp3-tibet-jpl-contentbucket-81rn23hp7ppf.s3.us-west-2.amazonaws.com/00312fde-8f89-4c5c-a0d6-ec2df678f316/00312fde-8f89-4c5c-a0d6-ec2df678f316.log'],
 'name': 'track106-n3',
 'priority': 100,
 'processing_times': [8558.771, 399.01],
 'request_time': '2023-08-25T03:38:19+00:00',
 'status_code': 'FAILED',
 'user_id': 'access_cloud_based_insar'}
```

As usual, here is the input GUNW to the above step:

https://hyp3-tibet-jpl-contentbucket-81rn23hp7ppf.s3.us-west-2.amazonaws.com/00312fde-8f89-4c5c-a0d6-ec2df678f316/S1-GUNW-A-R-106-tops-20170210_20170129-225948-00078W_00042N-PP-cde3-v3_0_0.nc

Hoping this can be quickly modified together.

### PR DESCRIPTION
**Pull request by cmarshak**
## Fixes
* Issue [#584](https://github.com/dbekaert/RAiDER/issues/584): failed Raider step function in hyp3 job submission when HRRR model times are not available (even within the valid model range) - to resolve, we check availability of files when delay workflow called with a) azimuth_grid_interpolation and b) input to workflow is GUNW. If weather model files are unavailable and the GUNW is on s3, do nothing to GUNW (i.e. do not add tropo delay) and exit successfully. If weather model files are unavailable and the GUNW is on local disk, raise `ValueError`
* Issue [#587](https://github.com/dbekaert/RAiDER/issues/587): similar to 584 except added here to the mix is control flow in RAiDER.py passes over numerous exceptions in workflow. This is fixed identically as above.

## Removed
* Removes `update` option (either `True` or `False`) from calcGUNW workflow which asks whether the GUNW should be updated or not. In existing code, it was not being used/applied, i.e. previous workflow always updated GUNW. Removed input arguments related from respective functions so that it can be updated later.

## Added
* Allow for Hyp3 GUNW workflow with HRRR (i.e. specifying a gunw path in s3) to successfully exit if any of the HRRR model times required for `azimuth-time-grid` interpolation (which is default interpolatin method) are not available when using bucket inputs (i.e. only on the cloud)
* For GUNW workflow, when model is HRRR, azimuth_time_grid interpolation used, and using a local GUNW, if requisite weather model files are not available for  raise a ValueError (before processing)
* Raise a value error if non-unique dates are given in the function `get_inverse_weights_for_dates` in `s1_azimuth_timing.py` 
* Added metadata provenance for each delay layer that is included in GUNW and the cube workflow generally in `calcDelays` including:
   * `model_times_used` - the weather models used and interpolated for the delay calculation
   * `interpolation_method` - whether `none`, `center_time`, or `azimuth_time_grid` methods were used - see description in `calcDelayGUNW`
   * `scene_center_time` - the center time in which the associated SAR image was acquired
* Stages GMAO data for GUNW testing of correct dataset update i.e. in the test `test_GUNW_dataset_update`.
* Stages HRRR data for `test_HRRR_ztd` test.

## Changed
* Get only 2 or 3 model times required for azimuth-time-interpolation (previously obtained all 3 as it was easier to implement) - this ensures slightly less failures associated with HRRR availability. Importantly, if a acquisition time occurs during a model time, then we order by distance to the reference time and how early it occurs (so earlier times come first if two times are equidistant to the aquisition time).
* Made test names in `test_GUNW.py` more descriptive
* Numpy docstrings and general linting to modified function including removing variables that were not being accessed
* hrrr_download to ensure that `hybrid` coordinate is obtained regardless how herbie returns datacubes and ensures test_HRRR_ztd passes consistently
   * Remove the command line call in `test_HRRR_ztd.py` and call using the python mock up of CLI for better error handling and data mocking.
* Return xarray.Dataset types for RAiDER.calcGUNW.tropo_gunw_slc and RAiDER.raider.calcDelayGUNW for easier inspection and testing

### CODE REVIEW

**Comment by cmarshak** [comment]
For relevant GUNWs where weather models are not available here is the new output from the PR (the GUNW is from #584) and the same will be true for #587.

```
raider.py ++process calcDelaysGUNW -f raider.py ++process calcDelaysGUNW -f S1-GUNW-A-R-106-tops-20170210_20170129-225948-00078W_00042N-PP-cde3-v3_0_0.nc -m HRRR -interp azimuth_time_grid -m HRRR -interp azimuth_time_grid
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-10 23:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-10 22:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-11 00:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
💔 Did not find ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-29 23:00 UTC F00
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-29 22:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-30 00:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
Traceback (most recent call last):
  File "/Users/cmarshak/mambaforge/envs/RAiDER/bin/raider.py", line 8, in <module>
    sys.exit(main())
  File "/Users/cmarshak/bekaert-team/RAiDER/tools/RAiDER/cli/__main__.py", line 42, in main
    process_entry_point.load()()
  File "/Users/cmarshak/bekaert-team/RAiDER/tools/RAiDER/cli/raider.py", line 609, in calcDelaysGUNW
    raise ValueError('The required data for interpolation is not available; returning None and not modifying GUNW dataset')
ValueError: The required HRRR data for time-grid interpolation is not available
```
and for S3 buckets
```
raider.py ++process calcDelaysGUNW --bucket s3://dummy/ --bucket-prefix S1-GUNW-A-R-106-tops-20170210_20170129-225948-00078W_00042N-PP-cde3-v3_0_0.nc -m HRRR -interp azimuth_time_grid -m HRRR -interp azimuth_time_grid
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-10 23:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-10 22:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Feb-11 00:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
💔 Did not find ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-29 23:00 UTC F00
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-29 22:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
✅ Found ┊ model=hrrr ┊ product=nat ┊ 2017-Jan-30 00:00 UTC F00 ┊ GRIB2 @ aws ┊ IDX @ aws
The required HRRR data for time-grid interpolation is not available; returning None and not modifying GUNW dataset
```

File: tools/RAiDER/cli/raider.py [hunk 1]
```diff
@@ -76,7 +77,7 @@ def read_template_file(fname):
             params[key] = {}
 
     # Parse the user-provided arguments
-    template = DEFAULT_DICT
+    template = DEFAULT_DICT.copy()
```
Comment (cmarshak) [inline tools/RAiDER/cli/raider.py:80]: This is an extremely important piece that @jhkennedy help me added that ensures tests pass. The `DEFAULT_DICT` is mutable and if its changed during runtime (i.e. during tests) there are mind boggling consequences (might be worth changing default dict to a constructor (e.g. `frozen` dataclasss so it's constant, but for now this resolves this issue). There is a fixture in `test/test_checkArgs.py` that loads this. After adding this, tests that were failing, passed.

File: tools/RAiDER/models/hrrr.py [hunk 2]
```diff
@@ -59,16 +68,21 @@ def download_hrrr_file(ll_bounds, DATE, out, model='hrrr', product='nat', fxx=0,
...
-            break
+    # Note order coord names are request for `test_HRRR_ztd` matters
+    # when both coord names are retreived by Herbie is ds_list possibly in
+    # Different orders on different machines; `hybrid` is what is expected for the test.
+    ds_list_filt_0 = [ds for ds in ds_list if 'hybrid' in ds._coord_names]
+    ds_list_filt_1 = [ds for ds in ds_list if 'isobaricInhPa' in ds._coord_names]
+    if ds_list_filt_0:
+        ds_out = ds_list_filt_0[0]
+        coord = 'hybrid'
+    #  I do not think that this coord name will result in successful processing nominally as variables are
```
Comment (jlmaurer) [inline tools/RAiDER/models/hrrr.py:79]: Change this comment to "TODO: check whether `isobaricInhPa` workflow will result in successfull processing"

File: tools/RAiDER/cli/raider.py [hunk 3]
```diff
@@ -410,12 +416,18 @@ def calcDelays(iargs=None):
         else:
             out_filename = w
 
+        # A dataset was returned by the above
+        # Dataset returned: Cube e.g. GUNW workflow
```
Comment (jlmaurer) [inline tools/RAiDER/cli/raider.py:420]: Are these comments needed?

**Review by jlmaurer — APPROVED** [review]
@cmarshak I think this looks good! I actually added a bunch of your updates in #585 before I read this PR. I just had a few minor comments (on your comments actually!); I think it will be best to merge this first then I will rebase #585 on top of it.

**Comment by jlmaurer** [comment]
@cmarshak I'm going to go ahead and merge this and will modify the comments in the next PR.

File: tools/RAiDER/cli/raider.py (hunk 3 above)
Comment (cmarshak) [reply tools/RAiDER/cli/raider.py:420]: `tropo_delay` has the [two different return types](https://github.com/dbekaert/RAiDER/blob/46da6a8ca231f6c99b9c43ee4632eccaf55067b9/tools/RAiDER/delay.py#L56-L57) and I was trying to figure out which of these two options the GUNW was using so I could add a breakpoint - would be nice to have a sign post. 

Can change language so long as going back it's clear that GUNW uses a particular branch.

**Pull request merged**